frontend/dockerfile/instructions/parse_test.go

```diff
@@ -140,6 +141,36 @@ func TestParseOptInterval(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestCommentsDetection(t *testing.T) {
+	dt := `# foo sets foo
+ARG foo=bar
+
+# base defines first stage
+FROM busybox AS base
+# this is irrelevant
+ARG foo
+# bar defines bar
+ARG bar
+`
+
+	ast, err := parser.Parse(bytes.NewBuffer([]byte(dt)))
+	require.NoError(t, err)
+
+	stages, meta, err := Parse(ast.AST)
+	require.NoError(t, err)
+
+	require.Equal(t, "defines first stage", stages[0].Comment)
```

Comment:
Why not include Key in Comment?

Reply:
The key can always be found from the object that contains the comment. It would work both ways: if a full comment is needed key can be added. And if the key was already in comment it could be trimmed if it is not needed.

In actual usage I'd think having full comment is less useful as you would want to show the typed parameters like the key or default value separately anyway so showing it also in the comment would be a duplication.